### PR TITLE
Review

### PR DESCRIPTION
**Pull request by GregTheGreek**
## Changes
- 
- 
-

### CODE REVIEW

File: chains/ethereum/chain.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	bridgeContract, err := createBridgeContract(cfg.bridgeContract, conn)
+	if err != nil {
+		return nil, err
+	}
+
+	erc20HandlerContract, err := createErc20HandlerContract(cfg.erc20HandlerContract, conn)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (GregTheGreek) [inline chains/ethereum/chain.go:102]: Why don't we initialize the generic contract?

File: chains/ethereum/connection.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	ctx       context.Context
+	conn      *ethclient.Client
+	signer    ethtypes.Signer
+	kp        *secp256k1.Keypair
+	nonceLock sync.Mutex
+	log       log15.Logger
+}
+
+func NewConnection(cfg *Config, kp *secp256k1.Keypair, log log15.Logger) *Connection {
+	signer := ethtypes.HomesteadSigner{}
```
Comment (GregTheGreek) [inline chains/ethereum/connection.go:44]: This will need to be configurable.

File: chains/ethereum/connection.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	}
+	return &ActiveSubscription{
+		ch:  ch,
+		sub: sub,
+	}, nil
+}
+
+// SubmitTx submits a transaction to the chain
+// It assumes the input data is an ethtypes.Transaction, marshalled as JSON
+func (c *Connection) SubmitTx(data []byte) error {
```
Comment (GregTheGreek) [inline chains/ethereum/connection.go:99]: I was thinking, could we capture the Nonce at this stage?
Furthermore, we might benefit from a queue, where we assign the Nonce as we fire them off.

File: chains/ethereum/deploy.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	centrifugeHandlerAddr, err := deployCentrifugeHandler(auth, client, bridgeAddr, deployAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	deployedContracts := DeployedContracts{bridgeAddr, relayerAddr, erc20HandlerAddr, erc721HandlerAddr, centrifugeHandlerAddr}
```
Comment (GregTheGreek) [inline chains/ethereum/deploy.go:81]: NIT: No need to assign you can just return

File: chains/ethereum/listener.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+			{sig.GetTopic()},
+		},
+	}
+	return query
+}
+
+// RegisterEventHandler creates a subscription for the provided event on the bridge bridgeContract.
+// Handler will be called for every instance of event.
+func (l *Listener) RegisterEventHandler(subscription EventSig, handler evtHandlerFn) error {
+	evt := subscription
```
Comment (GregTheGreek) [inline chains/ethereum/listener.go:94]: This seems unnecessary

File: chains/ethereum/writer.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+func (w *Writer) ResolveMessage(m msg.Message) bool {
+	w.log.Trace("Attempting to resolve message", "type", m.Type, "src", m.Source, "dst", m.Destination)
+
+	switch m.Type {
+	case msg.FungibleTransfer:
+		return w.createErc20DepositProposal(m)
+	case msg.NonFungibleTransfer:
+		panic("not yet implemented")
+	case msg.GenericTransfer:
+		return w.createGenericDepositProposal(m)
```
Comment (GregTheGreek) [inline chains/ethereum/writer.go:58]: I dont see the contract being set, I'm confused at the pattern being followed for handling generic transfers?

File: chains/ethereum/writer.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+func u32toBigInt(n uint32) *big.Int {
+	return big.NewInt(int64(n))
+}
+
+func byteSliceTo32Bytes(in []byte) [32]byte { //nolint:deadcode,unused
+	out := [32]byte{}
+	// Note: this is safe as copy uses the min length of the two slices
+	copy(out[:], in)
+	return out
+}
```
Comment (GregTheGreek) [inline chains/ethereum/writer.go:82]: It might make sense to push these out to a helper file

File: chains/ethereum/writer_methods.go [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+	msg "github.com/ChainSafe/ChainBridge/message"
+	"github.com/ChainSafe/log15"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+const VoteDepositProposalMethod = "voteDepositProposal"
+const ExecuteDepositMethod = "executeDepositProposal"
+
+func constructErc20ProposalData(amount, tokenId, recipient []byte) []byte {
+	var data []byte
```
Comment (GregTheGreek) [inline chains/ethereum/writer_methods.go:18]: Link to the solidity code where this is defined might be helpfull

File: chains/ethereum/chain.go (hunk 1 above)
Comment (ansermino) [reply chains/ethereum/chain.go:102]: Updated in #320 (#323)

File: chains/ethereum/connection.go (hunk 2 above)
Comment (ansermino) [reply chains/ethereum/connection.go:44]: Done in #327

File: chains/ethereum/connection.go (hunk 3 above)
Comment (ansermino) [reply chains/ethereum/connection.go:99]: To be removed, we don't ever use `SubmitTx` #323

File: chains/ethereum/writer.go (hunk 6 above)
Comment (ansermino) [reply chains/ethereum/writer.go:58]: All writes are to the bridge contract, updated in #320

File: chains/ethereum/writer.go (hunk 7 above)
Comment (ansermino) [reply chains/ethereum/writer.go:82]: Removed in #320

**Pull request merged**